Support for podcast 2.0 chapters

tonytamsf opened the pull request:
fix #4767 

Using this test feed https://feed.nashownotes.com/rss.xml and chapter data for testing http://34.117.70.159/http%3Afeed.nashownotes.comrss.xml/http:1352.noagendanotes.com


- This will only support the json chapter format based on this spec and example [link](https://github.com/Podcastindex-org/podcast-namespace/tree/main/chapters)
- Of the option attribues, we are only loading `startTime`, `title`, `img` and `url`, ignoring `toc`, `endTime` and `location`
- The chapter information will be fetched and cached when an episode is downloaded
- To test the caching of chapter data, download an episode, turn on airplane mode and play it to see if the chapters are loaded

## Some thoughts that need to be decided
~~- [x] When there are 1-3 different chapter sources, I am giving Podcast 2.0 a slight precedence, I am still using the existing chapter merging logic.~~
- [x] download a cached chapter when an episode is downloaded
~~- [ ] there is a note about privacy at the end [here](https://github.com/Podcastindex-org/podcast-namespace/blob/main/chapters/jsonChapters.md) - we should consider what we will do to protect user privacy~~

![Screen Shot 2022-01-02 at 12 16 01 AM](https://user-images.githubusercontent.com/149837/147870037-1b283bbb-d502-4bb6-a2e5-92fa0233ded6.png)

shombando commented:
Hi @tonytamsf I'm entirely new and unfamiliar but saw this PR pop-up while I was looking at my own. So take everything I saw with a *giant block* of salt :)

As a long-time user of Antennapod, here's how I'd expect it to behave: If I downloaded an episode, I'd expect chapters and it's relevant media to work but I wouldn't expect it to fetch that data every time I start playing the episode (because currently the audio is not fetched each time). However, if I were streaming an episode, I'd expect it to be "fresh" as of when I started the stream (since I expect the app to download the entire audio... I may be wrong about how it actually works, but just stating my expectations).

All that being said, it would be interesting to download the chapters JSON file and all the media linked from that file and store all that locally and rewrite the chapters JSON file to point to the local media that has been downloaded. Perhaps it makes sense to create an archive with all the content. Maybe even include the mp3 in the archive, so it's a self-contained podcast episode file?

Hope this wasn't too presumptuous and at least somewhat helpful. Excited to see all of this land in the app, thanks!

keunes commented:
Yeey. Thanks @tonytamsf! Really happy this is coming 🙂

Regarding **downloading chapter data**: I think we said during the Hackathon to store & keep the data as long as the audio file is downloaded, refresh it whenever playback is initiated (if internet connection is available, if not, rely on older data), and delete the chapter data whenever the audio file is deleted.

Where to store it, I leave to the experts :)

Regarding **privacy in context of refreshing data when starting playback**: I like their suggestion to prompt the user and ask if they trust the podcast. Of course this is only relevant with external chapter files.
I am not sure, however, if this should be asked for each individual podcast, or if we should just have a global switch. It might be a matter of trust in the podcast, as much as it might be general willingness to make known when you are listening episodes. (Also because I'm likely to only listen to podcasts I trust - if I don't trust the people/organisation behind a podcast, would I trust it's content?)

We might show a modal, asking if refreshing-on-play should be permitted for this podcast, for all podcasts, or never.

Either way it will be challenging to concisely explain the risk and the choices. But I guess we'll manage (if we want to offer this choice.

tonytamsf commented:
Thank you @keunes and @shombando !  Great suggestions

tonytamsf commented:
@keunes When there are 1-3 different chapter sources, I am not giving Podcast 2.0 precedence, I am still using the existing chapter merging logic. What do you think?

keunes commented:
> I wouldn't expect it to fetch that data every time I start playing the episode (because currently the audio is not fetched each time) [@shombando]

The issue is that while the audio is not likely to change after release (publishing in the feed, at which point the episode might be auto-downloaded), while the chapter info _is_. For example No Agenda has a community member that makes the chapters for all episodes after they have been released, so upon episode download chapter info is not available yet. And this flexibility was also one of the reasons for putting podcasting 2.0 chapter info in a separate file, rather than in the feed itself.

However it doesn't make much sense to download chapter info again if you just pause playback a minute because someone asks you something. So maybe there can be a mechanism to prevent refresh if playback was paused less than 15 minutes ago (alternatively: if chapter data was downloaded/refreshed shorter than an hour ago).

And regarding streaming: I guess podcast data should be kept until another episode is started playback of, not erase immediately upon pause.

ByteHamster commented:
> However it doesn't make much sense to download chapter info again if you just pause playback a minute 

We can probably rely on OkHttp's cache there. The cache should already be enabled globally (but disabled for feed refresh). Not 100% sure if the code in this PR needs to do something to use it.

> I am not sure, however, if this should be asked for each individual podcast, or if we should just have a global switch.

Not completely sure if it should be a setting at all but if this is, I think it should be a global switch without a popup dialog. I consider every single popup dialog very bad practice. Once you start adding them, they get more and more, making the app a pain to use

keunes commented:
> When there are 1-3 different chapter sources, I am not giving Podcast 2.0 precedence

Actually I would personally ignore other sources if PC2.0 is available as it's most likely to be up to date (and to avoid a long chapters list and duplicates that might not be picked up). Especially those baked into the file, as they are updated least easily (on podcaster's side, but also because we don't refresh audio files) thus most likely out of date.

But if it's decided to use all sources, at least PC 2.0 should take precedence when deduping (because it is more likely to be up to date, and holds more metadata).

ByteHamster commented:
I don't think any chapter format should take precedence. Merging should always be used. Podcasters do all kinds of weird stuff.

For example, one would think that the podlove chapters from the feed would be more up-to-date than the media files but there are podcasts that just update the media file. One would think it would be easier to add chapter images to the podlove chapters in the feed than to the media file but some podcasts only add images to the media file and not to the feed.

Publishers will make the same mess with PodcastIndex chapters.

tonytamsf commented:
> I don't think any chapter format should take precedence. Merging should always be used. Podcasters do all kinds of weird stuff.
> 
> For example, one would think that the podlove chapters from the feed would be more up-to-date than the media files but there are podcasts that just update the media file. One would think it would be easier to add chapter images to the podlove chapters in the feed than to the media file but some podcasts only add images to the media file and not to the feed.
> 
> Publishers will make the same mess with PodcastIndex chapters.

In this code, I use the same [merging logic](https://github.com/AntennaPod/AntennaPod/blob/7623100bee9d8f7a7d852eafebea15c89e4fb179/core/src/main/java/de/danoeh/antennapod/core/feed/ChapterMerger.java#L21) that has been there for a while and double merge, giving a slight edge to PodcaseIndex (only if the number of chapters are the same, AND the chapter startTimes are off by more than 1 second)

tonytamsf commented:
> But if it's decided to use all sources, at least PC 2.0 should take precedence when deduping (because it is more likely to be up to date, and holds more metadata).

Yup, PC2.0 gets a very slight edge if the number of chapters are the same, and there is a starTime discrepency.

tonytamsf commented:
@keunes I would love to have you test this feature for me?

keunes commented:
Cool! I'll download the debug and try it out today.

keunes commented:
Just tried it with No Agenda and Podcasting 2.0. A few comments:
* Chapter images sometimes load a bit slower in the overview. It would be great to have the are where the image will appear show some slight grey background colour (like the colour behind episode covers), so the transition is a bit less stark.
* For the following chapter in No Agenda: 1413: "Flurona" http://1413.noagendanotes.com no image is shown, while I would expect the episode image. 
```
{"title":"Zyphr","url":"https://www.nationandstate.com/2019/09/04/project-pogo-project-zyphr-it-doesnt-get-more-serious-than-these-two-black-ops/","img":"","startTime":739}
```
* I see you also process the chapter URLs - nice! In the chapter overview it'd be nice to strip `https(s)//` from the view, so there's more space to see the actual domain.
* Long-pressing the URL in the chapter overview doesn't do anything. Could a long-press be a share action? So that I can easily share it, save to my notes app, etc?
* The chapter URLs should probably also show directly in the player screen, but that's for another PR I guess. I'll think and make a mock-up of how/where it could be shown.

Thanks again!

ByteHamster commented:
I would do the UI changes (removing https, long-press url) in another PR because they are not specific to Podcastindex support

keunes commented:
Btw. Any particular reason we ignore `toc`?

keunes commented:
> I would do the UI changes (removing https, long-press url) in another PR because they are not specific to Podcastindex support

Aren't chapter URLs specific to the podcast namespace chapters? Haven't heard of them in the context of other chapters so assumed the line for the URL in the chapter list was added with this PR.

ByteHamster commented:
Chapter URLs are supported by most other standards as well (that's why I don't really like that they re-invented yet another chapter standard. There is nothing really new about it https://xkcd.com/927/)

ByteHamster commented:
> Btw. Any particular reason we ignore toc?

We currently ignore that for the other standards as well, so we can display the full overview of all chapter images etc in the chapter list

keunes commented:
> We currently ignore that for the other standards as well, so we can display the full overview of all chapter images etc in the chapter list

Ok. Will create a separate issue for that, then. Some lists are quite long so anything that can be hidden is welcome :)

Matth7878 commented:
I skimmed through but from what I understood : chapters will be downloaded if a connection is available (and I guess if using mobile data is allowed) unless they have been fetched less than 15 minutes or if user has turned off option to auto refresh chapters info.
If so I think there should be a way to trigger manually an update for chapters.

Nice that you're working on that @tonytamsf 👍

tonytamsf commented:
> If so I think there should be a way to trigger manually an update for chapters.

That is a great idea

tonytamsf commented:
> We can probably rely on OkHttp's cache there. The cache should already be enabled globally (but disabled for feed refresh). Not 100% sure if the code in this PR needs to do something to use it.

I am currently downloading a cached version of the chapter when the user download an episode. You can test by downloading an episode, turn on airplane mode and play the episode and you can see the chapter loaded

tonytamsf commented:
I think I am ready for another review @keunes and @ByteHamster - not sure if anything else is needed from me

keunes commented:
I don't think there's been a commit since my review, right? So I wouldn't know how else to test than what already done.

From my review comment there's just one point that I think might not be UI-related:
> * For the following chapter in No Agenda: 1413: "Flurona" http://1413.noagendanotes.com no image is shown, while I would expect the episode image. 
`{"title":"Zyphr","url":"https://www.nationandstate.com/2019/09/04/project-pogo-project-zyphr-it-doesnt-get-more-serious-than-these-two-black-ops/","img":"","startTime":739}`

tonytamsf commented:
> I don't think there's been a commit since my review, right? So I wouldn't know how else to test than what already done.

Ah, there have been 2 commits since your last review.  I should have tagged you

1) podcast 2.0 gets a slight preference when there are 3 types of chapters. 

2) i will cache the chapter when an episode is downloaded

tonytamsf commented:
> For the following chapter in No Agenda: 1413: "Flurona" http://1413.noagendanotes.com no image is shown, while I would expect the episode image.
> {"title":"Zyphr","url":"https://www.nationandstate.com/2019/09/04/project-pogo-project-zyphr-it-doesnt-get-more-serious-than-these-two-black-ops/","img":"","startTime":739}

 
@keunes if a chapter does not have an image, you would want me to show the default image from the episode?  Does the user interface preference matter? (episode vs podcast feed image?). This is not a consistent behavior when I play chapters (from the non Podcast 2.0 spec) where there are no images at all.

![Screen Shot 2022-01-12 at 10 49 42 PM](https://user-images.githubusercontent.com/149837/149279790-16078a1e-cc9b-422c-b11b-137a82ce373f.png)

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/feed/ChapterMerger.java:
```diff
@@ -30,14 +30,14 @@ private ChapterMerger() {
                 if (Math.abs(chapterTarget.getStart() - chapterOther.getStart()) > 1000) {
                     Log.e(TAG, "Chapter lists are too different. Cancelling merge.");
-                    return chapters1;
+                    return chapters2;
```
I recently changed the logic here. Instead of preferring the first or second one, it now prefers the one that has "richer" data. This causes a merge conflict here (sorry).

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/storage/DBUpgrader.java:
```diff
@@ -326,6 +326,11 @@ static void upgrade(final SQLiteDatabase db, final int oldVersion, final int new
                     + " ADD COLUMN " + PodDBAdapter.KEY_MINIMAL_DURATION_FILTER + " INTEGER DEFAULT -1");
         }
+
+        if (oldVersion < 2090000) {
```
Please use version `2060000` because 2.6.0 is the version this will first be released in (not 2.9.0)

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/util/ChapterUtils.java:
```diff
@@ -123,6 +138,60 @@ private static CountingInputStream openStream(Playable playable, Context context
         }
     }
 
+    public static List<Chapter> loadChaptersFromUrl(String url) {
```
The rest of the class does not do parsing. It just contains the logic for starting the parsers. I think at least the JSON parsing code here would better fit into its own class. Not sure where to put it, maybe the `:parser:feed` module?

tonytamsf replied:
I moved the code to `model/feed/PodcastIndexChapter.java`, let me know what you think

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/util/ChapterUtils.java:
```diff
@@ -123,6 +138,60 @@ private static CountingInputStream openStream(Playable playable, Context context
+                    String img = null;
+                    int startTime = 0;
+                    JSONObject jsonObject = objChapters.getJSONObject(i);
+                    if (jsonObject.opt("startTime") != null) {
+                        startTime = jsonObject.getInt("startTime");
+                    }
```
If you use the `optInt()`, `optString()`, etc methods here, you do not need all the `if` statements and can directly write most of them into the constructor arguments. That would save about 20 lines of code and make it easier to read.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/util/ChapterUtils.java:
```diff
@@ -123,6 +138,60 @@ private static CountingInputStream openStream(Playable playable, Context context
+                    }
+                    chapters.add(chapter);
+                }
+            } catch (JSONException e) {
+                e.printStackTrace();
+                return null;
+            }
+        } catch (IOException e) {
+            Log.d(TAG, "Error loading Chapter" + url + e.toString());
+            return null;
+        }
```
Could both catch clauses be merged? They don't seem to do different things.

ByteHamster commented on model/src/main/java/de/danoeh/antennapod/model/feed/FeedItem.java:
```diff
@@ -427,6 +432,14 @@ public void removeTag(String tag) {
         tags.remove(tag);
     }
 
+    public String getChapterUrl() {
```
Everywhere else you call it `podcastIndexChapterUrl`. I think it would be good to mention it here, too. Or does that make the method name too long?

ByteHamster commented on model/src/main/java/de/danoeh/antennapod/model/feed/PodcastIndexChapter.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public String toString() {
+        return "PodcastIndexChapter "
+                + getTitle()
+                + " "
+                + getStart()
+                + " "
+                + getLink()
+                + " "
+                + getImageUrl();
```
That uses a lot of vertical screen space

```suggestion
                + getTitle()
                + " " + getStart()
                + " " + getLink()
                + " " + getImageUrl();
```

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/util/ChapterUtils.java:
```diff
@@ -65,10 +72,18 @@ public static void loadChapters(Playable playable, Context context) {
                 chaptersFromDatabase = DBReader.loadChaptersOfFeedItem(feedMedia.getItem());
             }
+
+            if (! StringUtil.isBlank(feedMedia.getItem().getChapterUrl())) {
```
From the StringUtil documentation:
> A minimal String utility class. Designed for internal jsoup use only - the API and outcome may change without notice. 

I would use Android's `TextUtils.isEmpty`

ByteHamster reviewed (commented):
Thanks for working on this and sorry for the late reply. Doing a new release always stresses me out a bit and then I don't have the energy to review larger PRs.

I have only looked at the code for now but it looks pretty good to me. Left a few smaller comments.

keunes commented:
> if a chapter does not have an image, you would want me to show the default image from the episode? Does the user interface preference matter? (episode vs podcast feed image?).

Sorry, I was talking about the main player, not the chapter list. The chapter list can remain without image indeed, but the player should have an image at all times. It should always be the episode image (the preference/setting determines only what is shown in lists).

tonytamsf commented:
> Sorry, I was talking about the main player, not the chapter list. The chapter list can remain without image indeed, but the player should have an image at all times. It should always be the episode image (the preference/setting determines only what is shown in lists).

@keunes 
I'm not sure how my code could have changed the way episode images are displayed.  Could you give me a sample feed and episode?

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/util/ChapterUtils.java:
```diff
@@ -65,10 +69,18 @@ public static void loadChapters(Playable playable, Context context) {
         List<Chapter> chaptersFromMediaFile = ChapterUtils.loadChaptersFromMediaFile(playable, context);
-        List<Chapter> chapters = ChapterMerger.merge(chaptersFromDatabase, chaptersFromMediaFile);
+        List<Chapter> chaptersMergePhase1 = ChapterMerger.merge(chaptersFromDatabase, chaptersFromMediaFile);
+        // Very slight preference to PodcastIndex chapters if the length is the same
```
That's no longer the case - please remove the comment :)

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/util/ChapterUtils.java:
```diff
@@ -123,6 +135,23 @@ private static CountingInputStream openStream(Playable playable, Context context
+            if (response.body() == null) {
+                return null;
+            }
+            chapters = PodcastIndexChapter.parseChapters(response.body().toString());
+            return chapters;
```
You can save a line of code here ;)

```suggestion
            return PodcastIndexChapter.parseChapters(response.body().toString());
```

ByteHamster commented on model/src/main/java/de/danoeh/antennapod/model/feed/PodcastIndexChapter.java:
```diff
@@ -0,0 +1,67 @@
+                + " " + getImageUrl();
+    }
+
+    public static List<Chapter> parseChapters(String jsonStr) {
```
Hmm, the model should only be a data store, not do business logic like parsing. Please move the method to another class. Maybe in the `:parser:feed` module? It's not a perfect match there but starting a new module just for PodcastIndex chapter parsing would be a bit too much

ByteHamster reviewed (commented):
Just some small comments left. Everything else looks ready to be merged :)

keunes commented:
> I'm not sure how my code could have changed the way episode images are displayed. Could you give me a sample feed and episode?

I just wanted to give it another try and so I downloaded the PlayDebug version from the GH runner, but:
* The downloads never seem to finish - it completes the bar but states never change into completed, so I cannot press Play but only Cancel download
* Thus for testing I can only stream, and with http://feed.nashownotes.com/rss.xml episode 1426 I am not getting any chapters in the player, even though the episode has a podcast:chapter tag.

So I can't reproduce the original issue with the image not appearing.

ByteHamster commented:
> The downloads never seem to finish - it completes the bar but states never change into completed, so I cannot press Play but only Cancel download

Sorry for that. It was fixed on the `master` branch for the beta releases but not merged into `develop` yet. I now merged the fix into this PR.

> I am not getting any chapters in the player, even though the episode has a podcast:chapter tag.

There seems to be something wrong with this PR: I am getting the following error message on logcat:

> D/PodcastIndexChapter: Error loading Chapterorg.json.JSONException: Value okhttp3.internal.http.RealResponseBody@ce0f4bf of type java.lang.String cannot be converted to JSONObject

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/util/ChapterUtils.java:
```diff
@@ -123,6 +135,23 @@ private static CountingInputStream openStream(Playable playable, Context context
+            if (response.body() == null) {
+                return null;
+            }
+            chapters = PodcastIndexChapter.parseChapters(response.body().toString());
```
Does the body need to be closed? Maybe the `string` method closes it automatically, not sure about `toString`. Getting a StrictMode violation:

```
2022-02-19 14:25:17.429 28336-28364/de.danoeh.antennapod.debug D/StrictMode: StrictMode policy violation: android.os.strictmode.LeakedClosableViolation: A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
        at android.os.StrictMode$AndroidCloseGuardReporter.report(StrictMode.java:1924)
        at dalvik.system.CloseGuard.warnIfOpen(CloseGuard.java:303)
        at java.io.FileOutputStream.finalize(FileOutputStream.java:492)
        at java.lang.Daemons$FinalizerDaemon.doFinalize(Daemons.java:291)
        at java.lang.Daemons$FinalizerDaemon.runInternal(Daemons.java:278)
        at java.lang.Daemons$Daemon.run(Daemons.java:139)
        at java.lang.Thread.run(Thread.java:923)
     Caused by: java.lang.Throwable: Explicit termination method 'close' not called
        at dalvik.system.CloseGuard.openWithCallSite(CloseGuard.java:259)
        at dalvik.system.CloseGuard.open(CloseGuard.java:230)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:253)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:186)
        at okio.Okio.sink(Okio.java:188)
        at okhttp3.internal.io.FileSystem$1.sink(FileSystem.java:49)
        at okhttp3.internal.cache.DiskLruCache$Editor.newSink(DiskLruCache.java:900)
        at okhttp3.Cache$CacheRequestImpl.<init>(Cache.java:442)
        at okhttp3.Cache.put(Cache.java:254)
        at okhttp3.Cache$1.put(Cache.java:149)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:131)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:127)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at de.danoeh.antennapod.core.service.BasicAuthorizationInterceptor.intercept(BasicAuthorizationInterceptor.java:27)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:257)
        at okhttp3.RealCall.execute(RealCall.java:93)
        at de.danoeh.antennapod.core.util.ChapterUtils.loadChaptersFromUrl(ChapterUtils.java:142)
        at de.danoeh.antennapod.core.util.ChapterUtils.loadChapters(ChapterUtils.java:74)
        at de.danoeh.antennapod.core.service.playback.PlaybackServiceTaskManager.lambda$startChapterLoader$2$PlaybackServiceTaskManager(PlaybackServiceTaskManager.java:230)
        at de.danoeh.antennapod.core.service.playback.-$$Lambda$PlaybackServiceTaskManager$K4sCpoV4FY-NLB-eiSWRi0L-X_c.subscribe(Unknown Source:4)
        at io.reactivex.internal.operators.completable.CompletableCreate.subscribeActual(CompletableCreate.java:39)
        at io.reactivex.Completable.subscribe(Completable.java:2302)
        at io.reactivex.internal.operators.completable.CompletableSubscribeOn$SubscribeOnObserver.run(CompletableSubscribeOn.java:64)
        at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
        at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
        at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
```

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/util/ChapterUtils.java:
```diff
@@ -123,6 +135,23 @@ private static CountingInputStream openStream(Playable playable, Context context
+        } catch (IOException e) {
+            Log.d(TAG, "error loading data from " + url);
+        } finally {
+            return null;
```
Returning in a `finally` block is a bit weird because the `finally` block is executed after the `try` block and therefore after the actual return above. Please move the return out of the block.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/util/ChapterUtils.java:
```diff
@@ -123,6 +135,23 @@ private static CountingInputStream openStream(Playable playable, Context context
+            chapters = PodcastIndexChapter.parseChapters(response.body().toString());
+            return chapters;
+        } catch (IOException e) {
+            Log.d(TAG, "error loading data from " + url);
```
Please use `e.printStackTrace`, so that debug logs are useful for fixing potential bugs

ByteHamster commented on model/src/main/java/de/danoeh/antennapod/model/feed/PodcastIndexChapter.java:
```diff
@@ -0,0 +1,67 @@
+                return chapters;
+            }
+        } catch (JSONException e) {
+            Log.d(TAG, "Error loading Chapter" + e.toString());
```
Please use `e.printStackTrace`, so that debug logs are useful for fixing potential bugs

ByteHamster commented on model/src/main/java/de/danoeh/antennapod/model/feed/PodcastIndexChapter.java:
```diff
@@ -0,0 +1,67 @@
+        } catch (JSONException e) {
+            Log.d(TAG, "Error loading Chapter" + e.toString());
+        } finally {
+            return null; 
```
Please move `return` out of `finally`.

tonytamsf commented:
> > The downloads never seem to finish - it completes the bar but states never change into completed, so I cannot press Play but only Cancel download
> 
> Sorry for that. It was fixed on the `master` branch for the beta releases but not merged into `develop` yet. I now merged the fix into this PR.


Fixed now

> > I am not getting any chapters in the player, even though the episode has a podcast:chapter tag.
> 
> There seems to be something wrong with this PR: I am getting the following error message on logcat:
> 
> > D/PodcastIndexChapter: Error loading Chapterorg.json.JSONException: Value okhttp3.internal.http.RealResponseBody@ce0f4bf of type java.lang.String cannot be converted to JSONObject

Fixed

@keunes @ByteHamster I'm ready for another test and code review.  Thanks for your patience.

tonytamsf commented:
> > I'm not sure how my code could have changed the way episode images are displayed. Could you give me a sample feed and episode?
> 
> I just wanted to give it another try and so I downloaded the PlayDebug version from the GH runner, but:
> 
> * The downloads never seem to finish - it completes the bar but states never change into completed, so I cannot press Play but only Cancel download
> * Thus for testing I can only stream, and with http://feed.nashownotes.com/rss.xml episode 1426 I am not getting any chapters in the player, even though the episode has a podcast:chapter tag.
> 
> So I can't reproduce the original issue with the image not appearing.

@keunes I'm ready for another review

keunes commented:
Ok, I tested it again. Seems to work now 👍
Episode 1422 of No Agenda has the following chapter without image:
```
{"title":"New term: E.D.G.E.","url":"","img":"","startTime":9097},
```
And the episode image is displayed instead in the full player (while no image is used in chapter list).

ByteHamster commented:
Just tried it and it works great :+1: I could not get the "Download » Airplane mode » Have chapters" workflow to work, though. I pushed a commit that makes caching a bit more aggressive. That's also good for privacy :) Let me know what you think.

keunes commented:
> I pushed a commit that makes caching a bit more aggressive.

How long is the cache kept? Like, if I download the chapter now, and start playback in two weeks, is it then checked for updates?

tonytamsf commented:
> Just tried it and it works great 👍 I could not get the "Download » Airplane mode » Have chapters" workflow to work, though. I pushed a commit that makes caching a bit more aggressive. That's also good for privacy :) Let me know what you think.

Yup, looks good. Maybe my previous cache was larger than yours when I tried the airplane mode.

It is ready to merge when you are ready

@ByteHamster

tonytamsf commented:
> How long is the cache kept? Like, if I download the chapter now, and start playback in two weeks, is it then checked for updates?

I think it depends on if you downloaded other items into the cache (images, chapters) and overflow your cache.  It is not deterministic

Matth7878 commented:
Wouldn't it be possible to keep cached chapters as a long as episode is still downloaded (on device) or in queue or favorites if episode is not downloaded but streamed?

If not there is a risk that you won't be able to access chapters. Especially if it is an old episode kept in favorites.

ByteHamster commented:
>> How long is the cache kept? Like, if I download the chapter now, and start playback in two weeks, is it then checked for updates?
> 
> I think it depends on if you downloaded other items into the cache (images, chapters) and overflow your cache. It is not deterministic

Yeah, exactly. Maybe we could modify it slightly: When the cached version returns an empty list, try to connect to the server and see if there is something new.

> Wouldn't it be possible to keep cached chapters as a long as episode is still downloaded (on device) or in queue or favorites if episode is not downloaded but streamed?

That sounds like a lot of manual management and I think we decided against that at some point

keunes commented:
@ByteHamster Earlier this year I wrote the following [above](https://github.com/AntennaPod/AntennaPod/pull/5630#issuecomment-1003683260):
> Regarding downloading chapter data: I think we said during the Hackathon to store & keep the data as long as the audio file is downloaded, refresh it whenever playback is initiated (if internet connection is available, if not, rely on older data), and delete the chapter data whenever the audio file is deleted.
> Where to store it, I leave to the experts :)

And that's what I had in mind also still now actually.

ByteHamster commented:
> refresh it whenever playback is initiated

This would mean that podcast creators know when you start playback, even if you have downloaded the file. Not sure if we want that, from a privacy perspective.

Matth7878 commented:
If no auto refresh when starting an episode is there a way to manually update chapters?
I am thinking about when no chapters is initially available.

ByteHamster commented:
> If no auto refresh when starting an episode is there a way to manually update chapters?

Currently not, no (well, clearing the cache but that's not something users should worry about)

> I am thinking about when no chapters is initially available.

That's why I proposed to keep checking when there are no chapters but stop doing it as soon as there is data (for privacy)

keunes commented:
> Currently not, no (well, clearing the cache but that's not something users should worry about)

Maybe a button 'Refresh'/'Refresh data' could be placed at the bottom (left from Cancel)? Or, better, a refresh icon in the top-right, which requires some visual changes but probably is less intrusive.

> That's why I proposed to keep checking when there are no chapters but stop doing it as soon as there is data (for privacy)

What happens if there are chapters in the file? Does that count as 'as soon as there is data'?

Other point: should we place 'Cancel' with 'Close'? There isn't anything to cancel, really.

tonytamsf commented:
I think this works as expected when you are connected to the internet, the chapter data will update when the phone can load data.  I think we are over complicating it when as users we drill down too much into the details of the cache.

When the user have downloaded an episode, it is possible the chapter data from the 2.0 spec might not be available, but it is likely loaded already.

I don't think we need to expose a button or anything else to the user.

I propose we ship this feature as is and see if used notice anything unexpected.

ByteHamster commented:
> I don't think we need to expose a button or anything else to the user.

I agree

> I propose we ship this feature as is and see if users notice anything unexpected?

Merged :) Thanks for implementing this. Will be released in AntennaPod 2.6.0

keunes commented:
Thank you @tonytamsf!

tonytamsf commented:
Thanks for testing the feature @keunes

dcalano commented:
Just wanted to say thanks for the work on this! :)

tonytamsf commented:
> Just wanted to say thanks for the work on this! :)

Thanks for the shout out, I enjoy adding 2.0 support!  welcome to the AntennaPod community
